### ISSUE TITLE
Emitting an event once connections are created

### ISSUE THREAD

**Issue by amerishk** [oracle/node-oracledb#535]
Is there an event which is emitted once connections are created in a pool? We need to know in our scenario about the completion of connection creation task and perform an activity. This activity need to wait until all connections are created.
Node JS version is 0.10.40
Node-OracleDB version is 1.6.0

**Comment by sagiegurari** [comment]
if you wrap with simple-oracledb than you have such events.
in pool level and process level:

https://github.com/sagiegurari/simple-oracledb#event-connection-created-oracledb

https://github.com/sagiegurari/simple-oracledb#event-connection-created

**Comment by dmcghan** [comment]
@amerishk Can you tell us more about what you're doing? We're curious about whether synchronous events are sufficient for most use-cases.

> We need to know in our scenario about the completion of connection creation task and perform an activity.

What sort of activity? Do you need the connection that was created in the activity?

> This activity need to wait until all connections are created.

@cjbj Would need to verify this, but I suppose if you set the `poolIncrement` = `poolMax`, then when you get the first connection all of the connections will have been created.

**Comment by cjbj** [comment]
@dmcghan the answer is not in all cases, e.g. with DRCP.

@amerishk you'll need to tell us more if you want to find a solution.  E.g. do you want to know when (to use OCI terms) sessions or connections are created?

**Comment by amerishk** [comment]
@dmcghan / @cjbj , currently, we have created a JS file for connection pool as no direct JS method was found in version 1.6.0 of Node-Oracle Module. We cannot move to higher version as of now, which is a limitation at our end.  
We are using this JS (connection pool) to create connections and while running the unit test cases (gulp mochaTest) of our application, we are not getting connection object and test cases are failing. The reason of this error which we analyzed was the delay in connection pool creation (due to async nature). 
Hence we wanted to know about an event which gets emitted and can tell the mochaTest gulp task to start executing.

**Comment by sagiegurari** [comment]
I didn't fully understand the issue, the explanation is not that clear, but mocha does support async initialization before running the tests.
checkout mocha docs: https://mochajs.org/

from the docs (that is just 1 example):
````js
  beforeEach(function(done) {
    db.clear(function(err) {
      if (err) return done(err);
      db.save([tobi, loki, jane], done);
    });
  });
````

If I completely didn't understand the issue, than ignore.

**Comment by amerishk** [comment]
We have to execute IntegrationTest cases in our application. In order to run, we need to connect to Oracle DB for Test cases. 
As a boot script, we are initializing the connection pool and in test case, we getting a connection object from the pool. We are facing the issue in the test case, at the time test case executes pool object does not get created and fails while getting the connection from the null pool object.

Hence, we were looking for a solution where test cases run only when the connection pool is created successfully. We were thinking of having an event associated with the connection pool create completion and notify the test case to start.
Please provide inputs.

**Comment by sagiegurari** [comment]
this has nothing to do with oracledb and you don't need the events.
what you need is to define a 'before' function in mocha with a 'done' parameter so it will be async, and in that function to create the pool and only after the pool is created to call the done and than the tests will run and will have a valid existing pool.

**Comment by cjbj** [comment]
Thanks for digging into this @sagiegurari 

For background info @dmcghan has done some thinking about how events might be implemented but much more thought is needed and it's not clear how it would all fit, e.g. you shouldn't just asynchronously signal events that result in trying to use one connection for multiple tasks.  Since there are other projects that are more urgent, work on events has been deferred.  Suggestions are still welcome.

Maybe a synchronous `createPool()` would help a few users.  This has been suggested a few times.

**Comment by sagiegurari** [comment]
synchronous createPool() is great.
I think I wrote on some other issue that it would allow us to create a module like this:

````js
module.exports = oracledb.createPool(....);
````

and than just do in any other module

````js
var myPool = require('./my-pool');
````

**Comment by dmcghan** [comment]
@sagiegurari The driver now includes an [internal pool cache](https://github.com/oracle/node-oracledb/blob/master/doc/api.md#connpoolcache) so you wouldn't _have_ to create/include other modules.

**Comment by cjbj** [comment]
Closing - old issue.
